src/pandas_openscm/index_manipulation.py

```diff
@@ -714,3 +714,95 @@ def update_levels_from_other(
     res = pd.MultiIndex(levels=levels, codes=codes, names=names)
 
     return res
+
+
+def set_levels(
+    ini: pd.MultiIndex, levels_to_set: dict[str, Any | Collection[Any]]
+) -> pd.MultiIndex:
+    """
+    Set the levels of a MultiIndex to the provided values
+
+    Parameters
+    ----------
+    ini
+        Input MultiIndex
+
+    levels_to_set
+        Mapping of level names to values to set. If values is of type `Collection`,
+        it must be of the same length as the MultiIndex. If it is not a `Collection`,
+        it will be set to the same value for all levels.
+
+    Returns
+    -------
+    :
+        New MultiIndex with the levels set to the provided values
+
+    Raises
+    ------
+        TypeError
+            If `ini` is not a MultiIndex
+        ValueError
+            If the length of the values is not equal to the length of the index
+    """
+    # TODO mypy says this is unreachable
+    if not isinstance(ini, pd.MultiIndex):
+        msg = f"Expected MultiIndex, got {type(ini)}"
+        raise TypeError(msg)
```

Comment:
Ah yes that's because the type hint for `ini` is `pd.MultiIndex`. You'll probably need to push this check one layer higher into the function that operates on DataFrames (where the index could be just index, rather than a multi-index)

Reply:
Can be resolved I think?